cleanrl/ppo_atari_envpool_xla_jax.py

```diff
@@ -0,0 +1,447 @@
+            return loss, (pg_loss, v_loss, entropy_loss, jax.lax.stop_gradient(approx_kl))
+
+        ppo_loss_grad_fn = jax.value_and_grad(ppo_loss, has_aux=True)
+
+        # clipfracs = []
+        for _ in range(args.update_epochs):
+            key, subkey = jax.random.split(key)
+            b_inds = jax.random.permutation(subkey, args.batch_size, independent=True)
+            for start in range(0, args.batch_size, args.minibatch_size):
+                end = start + args.minibatch_size
+                mb_inds = b_inds[start:end]
+                (loss, (pg_loss, v_loss, entropy_loss, approx_kl)), grads = ppo_loss_grad_fn(
+                    agent_state.params,
+                    b_obs[mb_inds],
+                    b_actions[mb_inds],
+                    b_logprobs[mb_inds],
+                    b_advantages[mb_inds],
+                    b_returns[mb_inds],
+                )
+                agent_state = agent_state.apply_gradients(grads=grads)
+        return agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    next_obs = envs.reset()
+    next_done = np.zeros(args.num_envs)
+
+    @jax.jit
+    def rollout(agent_state, episode_stats, next_obs, next_done, storage, key, handle, global_step):
+        for step in range(0, args.num_steps):
+            global_step += 1 * args.num_envs
+            storage, action, key = get_action_and_value(agent_state, next_obs, next_done, storage, step, key)
+
+            # TRY NOT TO MODIFY: execute the game and log data.
+            episode_stats, handle, (next_obs, reward, next_done, _) = step_env_wrappeed(episode_stats, handle, action)
+            storage = storage.replace(rewards=storage.rewards.at[step].set(reward))
+        return agent_state, episode_stats, next_obs, next_done, storage, key, handle, global_step
+
+    for update in range(1, args.num_updates + 1):
+        update_time_start = time.time()
+        agent_state, episode_stats, next_obs, next_done, storage, key, handle, global_step = rollout(
+            agent_state, episode_stats, next_obs, next_done, storage, key, handle, global_step
+        )
+        storage = compute_gae(agent_state, next_obs, next_done, storage)
+        agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key = update_ppo(
+            agent_state,
+            storage,
+            key,
+        )
+        avg_episodic_return = np.mean(episode_stats.returned_episode_returns)
+        print(f"global_step={global_step}, avg_episodic_return={avg_episodic_return}")
+        print(episode_stats)
+
+        # TRY NOT TO MODIFY: record rewards for plotting purposes
+        writer.add_scalar("charts/avg_episodic_return", avg_episodic_return, global_step)
+        writer.add_scalar("charts/avg_episodic_length", np.mean(episode_stats.returned_episode_lengths), global_step)
```

Comment:
I'm getting the following error on these lines unless I do `jax.device_get`. 
```bash
NotImplementedError: Got <class 'jaxlib.xla_extension.DeviceArray'>, but numpy array, torch tensor, or caffe2 blob name are expected.
```
Even though the operation is `np.mean`, the return type on my system is `jaxlib.xla_extension.DeviceArray`.
Facing this on both `gpu` as well as `cpu`.

Reply:
Interesting... I havent' runinto this issue. I am gonna put this as part of the CI to avoid

Reply:
CI indeed catches this error. Thanks for reporting @kinalmehta. I have added the fix for it.